pkg/receive/multitsdb.go

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	terrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/shipper"
+	"github.com/thanos-io/thanos/pkg/store"
+	"golang.org/x/sync/errgroup"
+)
+
+type MultiTSDB struct {
+	dataDir         string
+	logger          log.Logger
+	reg             prometheus.Registerer
+	tsdbCfg         *tsdb.Options
+	tenantLabelName string
+	labels          labels.Labels
+	bucket          objstore.Bucket
+	upload          bool
+
+	mtx         *sync.RWMutex
+	dbs         map[string]*FlushableStorage
```

Comment:
Yeah my initial thought (and parts of the implementation) were that these are distinct concerns and may be individually optimizable (some of these at some point were slices instead of maps). But the further I got, the clearer it actually became that this is precisely our _tenant_ definition. So I'll adapt to make that more clear.

Reply:
Moved to a `tenant` struct. Let me know what you think.